File: src/game/object/Player.java
package game.object;

import engine.Camera;
import engine.Window;
import game.Assets;
import org.jetbrains.annotations.Nullable;

import static org.lwjgl.glfw.GLFW.*;

public class Player {
	private static final float RENDER_WIDTH = 16.0f;
	private static final float RENDER_HEIGHT = 32.0f;

	private static final float LEFT = 6.0f;
	private static final float RIGHT = 6.0f;

	private static final float DOWN = 0.0f;
	private static final float UP = 26.0f;

	private static final int SPEED_FLOOR = 1;

	public float x;
	public float y;
	public float yVel = 0;
	public float gravity = 300f;

	public int currency = 0;

	public int tilesPerSecond;
	public int jumpRange;

	public float calcJumpForce() {
		/*
		 * let h be max jump height, g be gravity, x be time
		 * equation for jumping parabola: h - gx^2
		 *
		 * solve for when height = 0, get x = -sqrt(h / g)
		 *
		 * the derivate is -2gx
		 *
		 * plug in -2g(-sqrt(h/g)) to get jump boost at t = 0 to achieve max height given gravity
		 *
		 * = 2g * sqrt(h / g)
		 */
		return 1.34f * gravity * (float)Math.sqrt(((jumpRange + 0.5f) * Level.TILE_SIZE) / gravity);
	}

	public Player(float x, float y, int tilesPerSecond, int jumpRange) {
		this.x = x;
		this.y = y;
		this.tilesPerSecond = tilesPerSecond;
		this.jumpRange = jumpRange;
	}

	public float centerX() {
		return x;
	}

	public float centerY() {
		return y + UP / 2.0f;
	}

	@Nullable
	private Collision findCollision(Collision... collisions) {
		var bestCollision = (Collision)null;

		for (var collision : collisions) {
			if (collision.hit) {
				if (bestCollision == null || collision.distance < bestCollision.distance) {
					bestCollision = collision;
				}
			}
		}

		return bestCollision;
	}

	public void update(Window window, Level level, float delta) { //20, 1620
		var left = x - LEFT;
		var right = x + RIGHT;
		var down = y + DOWN;
		var up = y + UP;

		var onGround = false;

		if (yVel <= 0) {
			var moveY = yVel * delta;

			var downCollision = findCollision(
				Collision.collideDown(level, left, down + moveY),
				Collision.collideDown(level, right, down + moveY)
			);

			if (downCollision != null) {
				onGround = true;
				y = downCollision.y + DOWN;

				if (yVel < 0) {
					yVel = 0;
				}
			}
		} else {
			var moveY = yVel * delta;

			var upCollision = findCollision(
				Collision.collideUp(level, left, up + moveY),
				Collision.collideUp(level, right, up + moveY)
			);

			if (upCollision != null) {
				y = upCollision.y - UP;

				if (yVel > 0) {
					yVel = 0;
				}
			}
		}

		if (onGround) {
			if (window.key(GLFW_KEY_SPACE) == GLFW_PRESS || window.key(GLFW_KEY_W) == GLFW_PRESS) {
				yVel = calcJumpForce();
			}
		} else {
			yVel -= gravity * delta;
		}

		if (window.key(GLFW_KEY_A) >= GLFW_PRESS) {
			var moveX = getSpeed() * delta;

			var leftCollision = findCollision(
				Collision.collideLeft(level, left - moveX, up),
				Collision.collideLeft(level, left - moveX, y),
				Collision.collideLeft(level, left - moveX, down + 1.0f)
			);

			if (leftCollision == null) {
				x -= moveX;
			} else {
				x = leftCollision.x + LEFT;
			}
		}

		if (window.key(GLFW_KEY_D) >= GLFW_PRESS) {
			var moveX = getSpeed() * delta;

			var rightCollision = findCollision(
				Collision.collideRight(level, right + moveX, up),
				Collision.collideRight(level, right + moveX, y),
				Collision.collideRight(level, right + moveX, down + 1.0f)
			);

			if (rightCollision == null) {
				x += moveX;
			} else {
				x = rightCollision.x - RIGHT;
			}
		}

		y += yVel * delta;
	}

	public boolean checkCollectCoins(Coin[] coins) {
		for (Coin coin : coins) {
			if (!coin.collected) {
				if (coin.distTo(centerX(), centerY()) <= 16.0f) {
					coin.collected = true;
					if (coin.tier == CoinLevel.gold) {
						currency += 1;

					} else if (coin.tier == CoinLevel.heavy) {
						currency += 3;

					} else if (coin.tier == CoinLevel.delta) {
						currency += 5;

					} else if (coin.tier == CoinLevel.pass) {
						return true;
					}
				}
			}
		}
		return false;
	}

	public float getSpeed() {
		var speed = (
			((tilesPerSecond - SPEED_FLOOR) * Level.TILE_SIZE) * (float)Math.pow(5.0f / 10.0f, (float)currency / 2.0f)
		) + (SPEED_FLOOR * Level.TILE_SIZE);

		if (speed < SPEED_FLOOR * Level.TILE_SIZE) speed = SPEED_FLOOR * Level.TILE_SIZE;

		return speed;
	}

	public void changeSpeed(int amount) {
		tilesPerSecond += amount;
	}

	public void render(Camera camera) {
		//
		//Assets.colorShader.enable().setMVP(camera.getMVP(
		//	x - RENDER_WIDTH / 2.0f,
		//	y,
		//	RENDER_WIDTH,
		//	RENDER_HEIGHT
		//)).uniform4f(0, 0.0f, 1.0f, 0.0f, 1.0f);
		//Assets.rect.render();

		Assets.playerTexture.bind();
		Assets.textureShader.enable().setMVP(camera.getMVP(
			x - RENDER_WIDTH / 2.0f,
			y,
			RENDER_WIDTH,
			RENDER_HEIGHT
		));
		Assets.rect.render();
	}
}

File: src/game/object/Player.kt
package game.`object`

import engine.Camera
import engine.Window
import game.Assets
import game.CCD
import game.Util
import game.Vector
import org.lwjgl.glfw.GLFW
import kotlin.math.PI
import kotlin.math.floor
import kotlin.math.pow
import kotlin.math.sqrt

class Player(var x: Float, var y: Float, var tilesPerSecond: Int, var jumpRange: Int) {
	var yVel = 0f
	var xVel = 0f
	var gravity = 300f
	var acceleration = 300f
	var currency = 0
	var onGround = false

	val floorAngle = 2.0f * PI.toFloat() / 7.0f

	fun calcJumpForce(): Float {
		/*
		 * let h be max jump height, g be gravity, x be time
		 * equation for jumping parabola: h - gx^2
		 *
		 * solve for when height = 0, get x = -sqrt(h / g)
		 *
		 * the derivative is -2gx
		 *
		 * plug in -2g(-sqrt(h/g)) to get jump boost at t = 0 to achieve max height given gravity
		 *
		 * = 2g * sqrt(h / g)
		 */
		return 2.0f * gravity * sqrt((jumpRange + 0.5f) * Level.TILE_SIZE / gravity)
	}

	fun centerX(): Float {
		return x
	}

	fun centerY(): Float {
		return y + UP / 2.0f
	}

	private fun findCollision(vararg collisions: Collision): Collision? {
		var bestCollision = null as Collision?
		for (collision in collisions) {
			if (collision.hit) {
				if (bestCollision == null || collision.distance < bestCollision.distance) {
					bestCollision = collision
				}
			}
		}
		return bestCollision
	}

	private fun angleIsFloor(angle: Float): Boolean {
		return angle in 0.0f..floorAngle || angle in PI - floorAngle..PI
			//angle in 0.0f - floorAngle..0.0f || angle in -PI..-PI + floorAngle
	}

	data class Hitbox(val left: Float, val right: Float, val down: Float, val up: Float)

	fun getHitbox(): Hitbox {
		return Hitbox(x - LEFT, x + RIGHT, y + DOWN, y + UP)
	}

	fun update(window: Window, level: Level, delta: Float) {
		/* control */

		var walking = false

		if (window.key(GLFW.GLFW_KEY_A) >= GLFW.GLFW_PRESS) {
			xVel -= acceleration * delta
			walking = true
		}
		if (window.key(GLFW.GLFW_KEY_D) >= GLFW.GLFW_PRESS) {
			xVel += acceleration * delta
			walking = true
		}

		val speed = speed
		if (xVel > speed) xVel = speed
		else if (xVel < -speed) xVel = -speed

		if (!walking) {
			if (xVel > 0) {
				xVel -= acceleration * delta
				if (xVel < 0) xVel = 0.0f
			} else if (xVel < 0) {
				xVel += acceleration * delta
				if (xVel > 0) xVel = 0.0f
			}
		}

		if (onGround && (
			window.key(GLFW.GLFW_KEY_SPACE) == GLFW.GLFW_PRESS || window.key(GLFW.GLFW_KEY_W) == GLFW.GLFW_PRESS
		)) {
			yVel = 160.0f//calcJumpForce()
		} else {
			yVel -= gravity * delta
		}

		/* collision */

		val move = CCD.Line(Vector(x, y), Vector(x + xVel * delta, y + yVel * delta))
		val lines = Collision2.collectLines(
			level,
			Collision2.getRange(move.start.x, move.end.x, 0.1f),
			Collision2.getRange(move.start.y, move.end.y, 0.1f),
			0.1f
		)
		val collision = CCD.findCollision(move, lines, 0.1f)

		if (collision == null) {
			x = move.end.x
			y = move.end.y
			onGround = false

		} else {
			x = collision.movePoint.x
			y = collision.movePoint.y

			collision.wallVectors.forEach { wallVector ->
				if (angleIsFloor(wallVector.angle())) {
					val velocityVector = Vector(xVel, 0.0f)
					velocityVector.setProject(wallVector)

					xVel = velocityVector.x
					yVel = velocityVector.y
					onGround = true

				} else {
					val velocityVector = Vector(xVel, yVel)
					velocityVector.setProject(wallVector)

					xVel = velocityVector.x
					yVel = velocityVector.y
				}
			}
		}

		if (y > 96.0f) print("$y,")
	}

	fun checkCollectCoins(coins: List<Coin>): Boolean {
		for (coin in coins) {
			if (!coin.collected) {
				if (coin.distTo(centerX(), centerY()) <= 16.0f) {
					coin.collected = true
					currency += when (coin.tier) {
						CoinLevel.GOLD -> 1
						CoinLevel.HEAVY -> 3
						CoinLevel.DELTA -> 5
						CoinLevel.PASS -> {
							return true
						}
					}
				}
			}
		}
		return false
	}

	fun checkDie(level: Level): Boolean {
		val (left, right, down, up) = getHitbox()

		return arrayOf(
			left to down,
			left to up,
			right to down,
			right to up,
		).any { (x, y) ->
			val subY = Util.mod(y, Level.TILE_SIZE)

			subY < 12.0f && level.access(
				floor(x / Level.TILE_SIZE).toInt(),
				floor(y / Level.TILE_SIZE).toInt()
			) == Level.TILE_LAVA
		}
	}

	val speed: Float
		get() {
			var speed = (tilesPerSecond - SPEED_FLOOR) * Level.TILE_SIZE * (5.0f / 10.0f)
				.pow((currency / 2.0f)) + SPEED_FLOOR * Level.TILE_SIZE
			if (speed < SPEED_FLOOR * Level.TILE_SIZE) speed = SPEED_FLOOR * Level.TILE_SIZE
			return speed
		}

	fun changeSpeed(amount: Int) {
		tilesPerSecond += amount
	}

	fun render(camera: Camera) {
		//
		//Assets.colorShader.enable().setMVP(camera.getMVP(
		//	x - RENDER_WIDTH / 2.0f,
		//	y,
		//	RENDER_WIDTH,
		//	RENDER_HEIGHT
		//)).uniform4f(0, 0.0f, 1.0f, 0.0f, 1.0f);
		//Assets.rect.render();
		Assets.playerTexture.bind()
		Assets.textureShader.enable().setMVP(
			camera.getMVP(
				x - RENDER_WIDTH / 2.0f,
				y,
				RENDER_WIDTH,
				RENDER_HEIGHT
			)
		)
		Assets.rect.render()
	}

	companion object {
		private const val RENDER_WIDTH = 16.0f
		private const val RENDER_HEIGHT = 32.0f
		private const val LEFT = 6.0f
		private const val RIGHT = 6.0f
		private const val DOWN = 0.0f
		private const val UP = 26.0f
		private const val SPEED_FLOOR = 1
	}
}
